proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2245 @@
+proposal's formalism). This would simplify the addition of ``0``.
+
+On the other hand, doing this loses the principle that linear data
+types and unrestricted data types are one and the same. And sacrifices
+much code reuse.
+
+Uniqueness instead of linearity
+-------------------------------
+
+Languages like Clean and Rust have a variant of linear types called
+uniqueness, or ownership, typing. This is a dual notion: instead of
+functions guaranteeing that they use their argument exactly once, and
+no restriction being imposed on the caller, with uniqueness type, the
+caller must guarantee that it has a non-aliased reference to a value,
+and the function has no restriction.
+
+Where uniqueness really shines, is for in-place mutation: the ``write``
+function can take a regular ``Array`` as an argument, it just needs to
+require that it is unique. Freezing is really easy: just drop the
+constraint that the ``Array`` is unique, it will never be writable
+again.
+
+With linear types, we need to have two types ``MArray`` (guaranteed
+unique) and ``Array``, just like in Haskell today. This is fine when
+we are freezing one array: just call ``freeze``. But what if we are
+freezing a list of arrays? Do we need to ``map freeze``? This is
+unfortunate (the problem is even more complicated if we start
+considering ``MArray (MArray a)``). It has a feel of ``Coercible``,
+but it does feel harder.
+
+On the other hand, other examples work better with linear types, such
+as fork-join parallelism. This is why Rust has a notion of so-called
+mutable borrowed reference, on which constraints are more akin to
+linear types (or rather, affine types, technically).
+
+Overall, uniqueness type system are significantly more complex to
+specify and implement than linear types systems such as this
+proposal's.
+
+Linearity-in-kinds
+------------------
+
+Instead of adding a type for linear function, we could classify types
+in two kinds: one of unrestricted types and one of linear
+types. A value of a linear type must be used in a linear fashion.
+
+This would get rid of the continuation of ``newMArray`` in the
+motivating ``MArray`` interface.
+
+The most natural way to do this, in Haskell, is to add a second
+parameter to ``TYPE`` (the first one is for levity polymorphism). So,
+ignoring the levity polymorphism, we would have ``TYPE 'One`` for linear
+types and ``TYPE 'Omega`` for unrestricted type. We get polymorphism by
+abstracting over the multiplicity.
+
+As interesting as it is, there is quite some complication associated
+to it. First, because of laziness, you can't have a function of type
+``(A :: TYPE 'One) -> (B :: TYPE 'Omega)`` (because you don't need to
+consume the result, hence you may not consume an argument that you
+have to consume). So what would be the type of the arrow? Something
+like ``forall (p :: Multiplicity) (q ⩽ p). p -> q -> q``. So we're
+introducing some kind of bounded polymorphism in our story. This is
```

Comment:
11. Given that this is unknown, what is your concrete proposal at this? Simply forbid linearity in pattern synonyms, presumably?

Reply:
We can always bail out this way, indeed. For all of these unknown cases.

Which means that they are not dangerous unknown. But we haven't thought about these enough to know what specification we like best. So I'd rather keep them as unknown for the time being. Does that make sense?

Reply:
Yes -- makes sense. This is something that can be refined by implementation.